torch/utils/_python_dispatch.py

```diff
@@ -143,20 +143,29 @@ class TorchDispatchMode(metaclass=TorchDispatchModeMeta):
     """
     # Force metaclass to generate constructor at the base of the hierarchy
     def __init__(self):
-        pass
+        self.ancestors: Set[TorchDispatchMode]
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
     def __enter__(self):
-        if hasattr(self, "inner"):
-            raise RuntimeError(f"{self} has already been used as a mode, please create and use a fresh version")
         old = _get_torch_dispatch_mode()
-        self.inner = old
+        if hasattr(self, "inner"):
+            assert hasattr(self, "ancestors")
+            if old is not None and old not in self.ancestors:
+                raise RuntimeError(f"{self} has already been used as a mode and is not valid in the current state, " +
+                                   "because the current mode is not its ancestor. Please use a fresh version")
```

Comment:
We've got these new error messages, but I don't see a test testing them. Should we add some?